MatrixSDK/Data/EventTimeline/Thread/MXThreadEventTimeline.swift

```diff
@@ -152,58 +156,55 @@ public class MXThreadEventTimeline: NSObject, MXEventTimeline {
-                if let eventsAfter = context.eventsAfter {
-                    events.append(contentsOf: eventsAfter)
-                }
-                
-                self.decryptEvents(events) {
-                    self.addEvent(context.event, direction: .forwards, fromStore: false)
 
-                    if let eventsBefore = context.eventsBefore {
-                        for event in eventsBefore {
-                            self.addEvent(event, direction: .backwards, fromStore: false)
-                        }
-                    }
+        //  We cannot use /context api here, like in a room timeline.
+        //  This is due to https://github.com/vector-im/element-meta/issues/150
+        //  So we're paginating with /relations api until we find the specified event.
+        //  Can be replaced by the old /context implementation when we have the fix in backend.
+
+        var eventFound: Bool = false
+        var paginationError: Error? = nil
+        let dispatchGroup = DispatchGroup()
+        var paginationBlock: () -> Void = {}
+        let result = MXHTTPOperation()
 
-                    if let eventsAfter = context.eventsAfter {
-                        for event in eventsAfter {
-                            self.addEvent(event, direction: .forwards, fromStore: false)
-                        }
+        paginationBlock = {
+            dispatchGroup.enter()
+            let newRequest = self.paginate(limit, direction: .backwards, onlyFromStore: false) { response in
+                switch response {
+                case .success:
+                    if self.store.eventExists(withEventId: initialEventId, inRoom: self.thread.roomId) {
+                        eventFound = true
+                    } else if self.canPaginate(.backwards) {
+                        paginationBlock()
+                    } else {
+                        paginationError = MXThreadEventTimelineError.initialEventNotFound
                     }
-                    
-                    self.backwardsPaginationToken = context.start
-                    self.forwardsPaginationToken = context.end
-                    //  TODO: We cannot paginate backward/forward on this point, because /relations api is not capable
-                    //  to paginate with these pagination tokens.
-                    self.hasReachedHomeServerBackwardsPaginationEnd = true
-                    self.hasReachedHomeServerForwardsPaginationEnd = true
-                    
-                    success()
+                case .failure(let error):
+                    paginationError = error
                 }
-            case .failure(let error):
-                MXLog.error("[MXThreadEventTimeline][\(self.timelineId)] resetPaginationAroundInitialEvent failed: \(error)")
-                failure(error)
+                dispatchGroup.leave()
+            }
+            result.mutate(to: newRequest)
+        }
+
+        paginationBlock()
```

Comment:
I'm curious as to why use a block here seeing as it isn't stored and is called immediately after being created? Otherwise LGTM 👍

Reply:
Because it's recursively getting next page, until the specified event found, so block used like a function here.

Reply:
Oh, ignore me! I completely missed that call inside was recursive on itself and not on the method we are already in 🤦‍♂️